src/Compilers/CSharp/Test/Semantic/Semantics/IteratorTests.cs

```diff
@@ -106,6 +106,182 @@ IEnumerable<int> I()
                 );
         }
 
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/72443")]
+        public void YieldInLock_Async()
+        {
+            var source = """
+                using System;
+                using System.Collections.Generic;
+                using System.Threading;
+                using System.Threading.Tasks;
+
+                try
+                {
+                    await new C().ProcessValueAsync();
+                }
+                catch (SynchronizationLockException e)
+                {
+                    Console.WriteLine(e.Message);
+                }
+
+                public class C
+                {
+                    public async Task ProcessValueAsync()
+                    {
+                        await foreach (int item in GetValuesAsync())
+                        {
+                            await Task.Yield();
+                            Console.Write(item);
+                        }
+                    }
+
+                    private async IAsyncEnumerable<int> GetValuesAsync()
+                    {
+                        await Task.Yield();
+                        lock (this)
+                        {
+                            for (int i = 0; i < 10; i++)
+                            {
+                                yield return i;
+
+                                if (i == 3)
+                                {
+                                    yield break;
+                                }
+                            }
+                        }
+                    }
+                }
+                """ + AsyncStreamsTypes;
+
+            var expectedOutput = ExecutionConditionUtil.IsDesktop
+                ? null
+                : "0123Object synchronization method was called from an unsynchronized block of code.";
```

Comment:
>"0123Object synchronization method was called from an unsynchronized block of code." [](http://example.com/codeflow?start=18&length=84)

I do not think we should be testing this specific runtime exception behavior, this isn't an exception from a specific API or an exception explicitly thrown by compiler generated code. Consider not validating the output at all.